### ISSUE TITLE
memory leak when refreshing webpages

### ISSUE THREAD

**Issue by ghost** [qutebrowser/qutebrowser#6141]
**Version info**:

```
qutebrowser v2.0.2
Git commit:
Backend: QtWebEngine (Chromium 83.0.4103.122)
Qt: 5.15.2

CPython: 3.8.7
PyQt: 5.15.2

sip: 5.4.0
colorama: no
jinja2: 2.11.2
pygments: 2.7.2
yaml: 5.3.1
adblock: 0.4.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebEngine: 5.15.2
PyQt5.QtWebKitWidgets: no
pdf.js: 2.6.347 (/nix/store/ysdvxa0qdkvd48fd3zswpmyspblzbhvw-pdfjs-2.6.347/build/pdf.js)
sqlite: 3.34.1
QtNetwork SSL: OpenSSL 1.1.1i  8 Dec 2020
```

**Does the bug happen if you start with `--temp-basedir`?**:
no

**Description**

While doing web development with qutebrowser the python process started using more and more memory, ending up at 4GB RES.

**How to reproduce**

As mentioned above, this only worked with my profile, but it may also be reproducible with other (non-empty) profiles.
Maybe some of the steps can be ommitted, but this is what worked for me:

- open htop, watch the qutebrowser python process row, RES column 
- open qutebrowser
- go to https://petabyte.dev
- open devtools
- continously press Ctrl+F5
- watch the memory usage climb

**Comment by The-Compiler** [comment]
So what I see here is that the memory usage rises (by a MB all couple of seconds or so), but goes down significantly after stopping to press Ctrl+F5. It then seems to stay at a higher level than initially though.

However, I see the exact same behavior in Chromium (with the difference that you can't hold Ctrl+F5, but holding Ctrl and smashing F5 works). Do you see the same?

Finally, you say "in qutebrowser 2.0.x" - does that imply this wasn't the case with older qutebrowser (but same Qt) versions?

**Comment by The-Compiler** [comment]
@petabyteboy Any update on this?

**Comment by petabyteboy** [comment]
I can still reproduce this, now even with `--temp-basedir`:

https://user-images.githubusercontent.com/3250809/107850218-0d2ce200-6e01-11eb-9195-170d5d28d3f8.mp4

I didn't get around to testing with chromium yet. Will do that later today.

> Finally, you say "in qutebrowser 2.0.x" - does that imply this wasn't the case with older qutebrowser (but same Qt) versions?

At least I didn't notice it until shortly after the update. I will try to reproduce with older qutebrowser versions.

**Comment by petabyteboy** [comment]
I can also reproduce it in a fresh Arch Linux installation in a VM ([version info](https://termbin.com/ei30)), without opening my website and without  opening the devtools:

https://user-images.githubusercontent.com/3250809/107851231-be834600-6e08-11eb-9d3d-677c3d35144d.mp4

The same does not happen with Chromium 83. It quickly settles at around 180MiB RES and always goes back to idle usage even after hammering it with reloads for 4 minutes.

**Comment by The-Compiler** [comment]
Thanks for the detailed update! I'll have a closer look as soon as I'm able to, but I suspect there's little qutebrowser can do. Could you try to reproduce with KDE Falkon (another browser using QtWebEngine) as well?

**Comment by petabyteboy** [comment]
> Thanks for the detailed update! I'll have a closer look as soon as I'm able to, but I suspect there's little qutebrowser can do. Could you try to reproduce with KDE Falkon (another browser using QtWebEngine) as well?

Great idea! Indeed, Falkon has the same issue. Can you take care of reporting the bug upstream?

**Comment by The-Compiler** [comment]
Only if I can reliably reproduce it - I'll retry later. If not, you will have to report it yourself, as I can't provide any information for follow-up questions and such.

**Comment by The-Compiler** [comment]
Hm, my results are still pretty inconclusive.

With both qutebrowser and Falkon, I think I see a similar behavior as you describe. The RAM usage climbs while reloading, then falls a bit when stopping to reload, but it then rests at a higher level very time.

However, with QtWebEngine's `simplebrowser` example, as well as qutebrowser's `testbrowser_webengine.py`, I *don't* see the issue...

Patch for the reloads:

```diff
diff --git i/scripts/testbrowser/testbrowser_webengine.py w/scripts/testbrowser/testbrowser_webengine.py
index 73fa6828e..f08e15046 100755
--- i/scripts/testbrowser/testbrowser_webengine.py
+++ w/scripts/testbrowser/testbrowser_webengine.py
@@ -23,7 +23,7 @@
 import sys
 import argparse
 
-from PyQt5.QtCore import QUrl
+from PyQt5.QtCore import QUrl, QTimer
 from PyQt5.QtWidgets import QApplication
 from PyQt5.QtWebEngineWidgets import QWebEngineView
 
@@ -48,4 +48,9 @@ def parse_args():
     wv.load(QUrl.fromUserInput(args.url))
     wv.show()
 
+    timer = QTimer()
+    timer.setInterval(100)
+    timer.timeout.connect(wv.reload)
+    timer.start()
+
     app.exec()

```

So I'm guessing there must be some QtWebEngine-feature or something which triggers this, and both qutebrowser and Falkon use it, but the simplebrowser example does not. I have no idea what that could be, though...

**Comment by dindon-sournois** [comment]
I'd to see if I can help as I also have the issue, forcing me to restart qutebrowser once in a while. What kind of reproducer would you need @The-Compiler ?

What I do to provoke the issue with a clear config / cache is open 5 or 6 search engine tabs, make a macro that will go over each tab and refresh them. Then I launch the macro maybe 100 or 200 times and in 10 seconds the memory usage rest at 5MB more than initially. I know this sounds like an extreme use of refresh but it's what I do to reproduce an issue that arises after hours of usage.

I'm not sure I understand what your patched reload test does but by putting the interval to 10 I was able to observe an ever increasing memory usage.

**Comment by The-Compiler** [comment]
See my last comment above. I haven't gotten around to dig into this in more detail yet. Also see #1476 FWIW.

**Comment by markstos** [comment]
One way to simplify the test case would be to try reloading an even simpler web page. For example, instead of Duck Duck Go, try reloading "about:blank" or some other extremely simply HTML page.  Once the simplest web page that triggers it has been found, review list of API calls that Qutebrowser makes into QtWebEngine. Subtract the calls that are also made by Simplebrowser. Either the leaking calls are in the list of methods remaining or they are elsewhere in the python code.

**Comment by markstos** [comment]
As a funder for the last several months, the memory leaks in Qutebrowser are my number one priority. I have have a 16 GB laptop and now have a near-daily ritual of having my desktop come to a crawl and running `killall qutebrowser`, and then watching from my memory use drop from about 16 GB to 2 GB. It's so bad so I've switched  to using Electron for Slack and Spotify to stabilize my memory use. 😂 I really like Qutebrowser but will be exploring alternatives and considering dropping my funding if memory use can't be stabilized. Neither Firefox or Chromium required such frequent restarts.

**Comment by The-Compiler** [comment]
@markstos Are you on Wayland? If so, this issue is probably irrelevant, see #1476 instead.

**Comment by markstos** [comment]
@The-Compiler Yes, I'm using Wayland. I'll check out the other issue. Thanks.

**Comment by toofar** [comment]
I was going to post this in the other memory usage issue but this one seems more relevant. There was a leak fixed in webengine that occurred on every network request. So consistent with reloading a page a bunch https://bugreports.qt.io/browse/QTBUG-96928
The fix is in QtWebEngine 5.15.7 (git tag v5.15.7-lts which is apparently in arch since two weeks ago) so hopefully the problem goes away on that version. Although if the reproduction wasn't consistent, maybe not.

**Comment by toofar** [comment]
I'll close this with the assumption the https://bugreports.qt.io/browse/QTBUG-96928 fixed it.
